### PR TITLE
Make RestException pickleable

### PR DESCRIPTION
**Pull request by WeichenXu123**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/WeichenXu123/mlflow/pull/10936?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10936/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10936
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Make MlflowException pickleable, reason:
Some Ray program wants to log data to MLflow server in remote Ray tasks, then, if remote Ray task raises exception, Ray framework tries to pickle the exception to client and then un-pickle it (so that Ray can show exception details in client side). But if a MLflowException is raised in remote Ray task side, unpickling the MLflow exception will fail and causes some unreadable error message like:
```
Traceback (most recent call last):
  File "/local_disk0/.ephemeral_nfs/envs/pythonEnv-13bb31d1-c099-4935-b303-67bb18adcb8d/lib/python3.10/site-packages/ray/_private/serialization.py", line 404, in deserialize_objects
  obj = self._deserialize_object(data, metadata, object_ref)
  File "/local_disk0/.ephemeral_nfs/envs/pythonEnv-13bb31d1-c099-4935-b303-67bb18adcb8d/lib/python3.10/site-packages/ray/_private/serialization.py", line 293, in _deserialize_object
    return RayError.from_bytes(obj)
  File "/local_disk0/.ephemeral_nfs/envs/pythonEnv-13bb31d1-c099-4935-b303-67bb18adcb8d/lib/python3.10/site-packages/ray/exceptions.py", line 40, in from_bytes
    return RayError.from_ray_exception(ray_exception)
  File "/local_disk0/.ephemeral_nfs/envs/pythonEnv-13bb31d1-c099-4935-b303-67bb18adcb8d/lib/python3.10/site-packages/ray/exceptions.py", line 49, in from_ray_exception
    raise RuntimeError(msg) from e
RuntimeError: Failed to unpickle serialized exception
RaySystemError: System error: Failed to unpickle serialized exception
traceback: Traceback (most recent call last):
  File "/local_disk0/.ephemeral_nfs/envs/pythonEnv-13bb31d1-c099-4935-b303-67bb18adcb8d/lib/python3.10/site-packages/ray/exceptions.py", line 46, in from_ray_exception
    return pickle.loads(ray_exception.serialized_exception)
  File "/local_disk0/.ephemeral_nfs/envs/pythonEnv-13bb31d1-c099-4935-b303-67bb18adcb8d/lib/python3.10/site-packages/mlflow/exceptions.py", line 117, in __init__
    error_code = json.get("error_code", ErrorCode.Name(INTERNAL_ERROR))
AttributeError: 'str' object has no attribute 'get'
```

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/exceptions.py [hunk 1]
```diff
@@ -67,6 +67,8 @@ class MlflowException(Exception):
     The error message associated with this exception may be exposed to clients in HTTP responses
     for debugging purposes. If the error text is sensitive, raise a generic `Exception` object
     instead.
+
+    Note: Instance of `MlflowException` is pickleable, see `__getstate__` and `__setstate__`
```
Comment (harupy) [inline mlflow/exceptions.py:71]: `MlflowException` is already picklable:

```
>>> import mlflow
>>> import pickle
>>> pickle.dumps(mlflow.MlflowException("hello"))
```

Comment (WeichenXu123) [reply mlflow/exceptions.py:71]: No. when `pickle.load` it fails.

Comment (WeichenXu123) [reply mlflow/exceptions.py:71]: You can repro it by the following code:

```
import mlflow.exceptions
import pickle
v1 = mlflow.exceptions.RestException({"error_code": "INTERNAL_ERROR", "message": "abc"})
s1 = pickle.dumps(v1)
v2 = pickle.loads(s1)  # error occurs
```

Error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/weichen.xu/work/mlflow/mlflow/exceptions.py", line 117, in __init__
    error_code = json.get("error_code", ErrorCode.Name(INTERNAL_ERROR))
AttributeError: 'str' object has no attribute 'get'
```

**Comment by WeichenXu123** [comment]
@mlflow-automation autoformat

File: tests/test_exceptions.py [hunk 2]
```diff
@@ -50,3 +52,12 @@ def test_rest_exception():
     deserialized_rest_exception = RestException(json.loads(json_exception))
     assert deserialized_rest_exception.error_code == "RESOURCE_ALREADY_EXISTS"
     assert "test" in deserialized_rest_exception.message
+
+
+def test_test_rest_exception_pickleable():
+    v1 = mlflow.exceptions.RestException({"error_code": "INTERNAL_ERROR", "message": "abc"})
```
Comment (harupy) [inline tests/test_exceptions.py:58]:
```suggestion
    v1 = RestException({"error_code": "INTERNAL_ERROR", "message": "abc"})
```

File: tests/test_exceptions.py [hunk 3]
```diff
@@ -50,3 +52,12 @@ def test_rest_exception():
...
+    s1 = pickle.dumps(v1)
+    v2 = pickle.loads(s1)
```
Comment (harupy) [inline tests/test_exceptions.py:60]:
```suggestion
    v2 = pickle.loads(pickle.dumps(v1))
```

File: mlflow/exceptions.py (hunk 1 above)
Comment (harupy) [inline mlflow/exceptions.py:71]: Where is `__getstate__` and `__setstate__`?

File: mlflow/exceptions.py [hunk 4]
```diff
@@ -122,6 +124,9 @@ def __init__(self, json):
         super().__init__(message, error_code=ErrorCode.Value(error_code))
         self.json = json
 
+    def __reduce__(self):
+        return type(self), (self.json,)
```
Comment (harupy) [inline mlflow/exceptions.py]: Let's add a comment. It's not obvious why we need this.

File: tests/test_exceptions.py [hunk 5]
```diff
@@ -50,3 +52,12 @@ def test_rest_exception():
     deserialized_rest_exception = RestException(json.loads(json_exception))
     assert deserialized_rest_exception.error_code == "RESOURCE_ALREADY_EXISTS"
     assert "test" in deserialized_rest_exception.message
+
+
+def test_test_rest_exception_pickleable():
```
Comment (harupy) [inline tests/test_exceptions.py:57]:
```suggestion
def test_rest_exception_ picklable():
```

File: mlflow/exceptions.py (hunk 1 above)
Comment (WeichenXu123) [reply mlflow/exceptions.py]: Removed. We only need to override `__reduce__`, if it has `__reduce__`, `__getstate__` and `__setstate__` is not used.

File: mlflow/exceptions.py [hunk 6]
```diff
@@ -122,6 +122,12 @@ def __init__(self, json):
         super().__init__(message, error_code=ErrorCode.Value(error_code))
         self.json = json
 
+    def __reduce__(self):
+        """
+        Overriding `__reduce__` to make `RestException` instance pickle-able.
+        """
+        return type(self), (self.json,)
```
Comment (harupy) [inline mlflow/exceptions.py:129]:
```suggestion
        return RestException, (self.json,)
```

File: tests/test_exceptions.py [hunk 7]
```diff
@@ -50,3 +52,11 @@ def test_rest_exception():
     deserialized_rest_exception = RestException(json.loads(json_exception))
     assert deserialized_rest_exception.error_code == "RESOURCE_ALREADY_EXISTS"
     assert "test" in deserialized_rest_exception.message
+
+
+def test_rest_exception_pickleable():
+    e1 = mlflow.exceptions.RestException({"error_code": "INTERNAL_ERROR", "message": "abc"})
```
Comment (harupy) [inline tests/test_exceptions.py:58]:
```suggestion
    e1 = RestException({"error_code": "INTERNAL_ERROR", "message": "abc"})
```

File: tests/test_exceptions.py [hunk 8]
```diff
@@ -1,5 +1,7 @@
 import json
+import pickle
 
+import mlflow
```
Comment (harupy) [inline tests/test_exceptions.py:4]:
```suggestion
```

**Comment by WeichenXu123** [comment]
@harupy Do you have any new comments ? :)

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by davidxia** [comment]
Is this released yet? If so, which version is it in?

If not, any idea when it'll be released?

**Comment by WeichenXu123** [comment]
@davidxia It will be included in MLflow 2.11 , which will be released soon.

**Pull request merged**